src/satosa/satosa_config.py

```diff
@@ -30,11 +31,12 @@ def __init__(self, config):
         :param config: Can be a file path or a dictionary
         :return: A verified SATOSAConfig
         """
-        parsers = [self._load_dict, self._load_yaml]
-        for parser in parsers:
-            self._config = parser(config)
-            if self._config is not None:
-                break
+        self._config = self._parse_config(config)
+
+        if not self._config:
+            raise SATOSAConfigurationError(
+                "Missing configuration or unknown format"
+            )
```

Comment:
Should this be part of `_parse_config`?

Reply:
Nope. This statement used to be in the beginning of `_verify_dict` which was passed the `self._config` as an argument. It was not used as a generic error, only to state the failure of the initial config parsing

Reply:
Let me rephrase the question. 
Should the check `if not self._config: raise error()` be inside `_parse_config`?

Reply:
Nope. `_parse_config` parses the argument `config`. It could as well be a staticmethod or function if the `_load_dict` and `_load_yaml` where not instance methods (and they could as well be external functions, there is no need to be coupled inside the SATOSAConfig object

Reply:
This is not answering what I am asking. Whether those are functions or methods is irrelevant to the question. What I am asking is _who is responsible for checking if the result of an action is an error_ (you do not need to think about this particular case at all.)

You have two choices, the first is what is being done now - the caller is checking for errors

```
def foo(x):
    value = bar(x)
    if not value:
        generate_error()
    do_something_with(value)

def bar(x):
    value = get_value(x)
    return value
```

and a second choice, where the callee is checking for the error

```
def foo(x):
    value = bar(x)
    do_something_with(value)

def bar(x):
    value = get_value(x)
    if not value:
        generate_error()
    return value
```

How can we compare those two and choose one over the other?

Reply:
It depends is the answer. It depends on what kind of error is this. Is it an evaluation error or it is a logic error? `bar` could throw errors if they have to do with the processing of `x` but if there is a logical error with the processing of `x`, should `bar` know the logic of `foo` and throw the error? I would say no

Reply:
OK, so what we do here is we parse the `config`. And parsing fails - none of the parsers can parse the given thing. By what you said, `_parse_config` should then throw the error, because it failed to do its job. The return value of `None` is not a _logical_ error, it is a _parsing_ error; a return value sentinel that signifies the failure of the operation. For this specific operation, an evaluation error would be a config that was parsed but has inconsistent values for the different options it holds, ie it would be a validation error.

Reply:
Indeed in this case, it is a parsing error and the parsing function should throw this